controller/lib/src/external_api/workload/controller.rs

```diff
@@ -1,46 +1,168 @@
-        .service(get_workloads)
-        .service(get_workload)
-        .service(put_workload)
-        .service(patch_workload)
-        .service(delete_workload)
-}
+use crate::external_api::interface::ActixAppState;
 
-#[get("/")]
-pub async fn get_workloads() -> impl Responder {
-    service::get_workloads(&["1".to_string(), "2".to_string(), "3".to_string()]);
-    HttpResponse::Ok().body("workload")
-}
+use super::service::WorkloadService;
+use super::{model::Pagination, model::WorkloadDTO};
+use actix_web::http::StatusCode;
+use actix_web::{web, HttpResponse, Responder, Scope};
+pub struct WorkloadController {}
 
-#[get("/{workload_id}")]
-pub async fn get_workload(workload_id: web::Path<String>) -> impl Responder {
-    service::get_workloads(&[workload_id.to_string()]);
-    HttpResponse::Ok().body(format!("workload {}", workload_id))
-}
+impl WorkloadController {
+    pub fn services(&self) -> Scope {
+        web::scope("/workload")
+            .service(
+                web::resource("/{namespace}/{workload_id}")
+                    .route(web::delete().to(WorkloadController::delete_workload))
+                    .route(web::get().to(WorkloadController::workload))
+                    .route(web::patch().to(WorkloadController::patch_workload)),
+            )
+            .service(
+                web::resource("/{namespace}")
+                    .route(web::put().to(WorkloadController::put_workload))
+                    .route(web::get().to(WorkloadController::get_all_workloads)),
+            )
+    }
 
-#[put("/")]
-pub async fn put_workload(body: web::Json<WorkloadInfo>) -> impl Responder {
-    let workload_info = body.into_inner();
-    service::create_workload(&workload_info);
-    HttpResponse::Ok().body("put_workload")
-}
+    /// It gets a workload from etcd, and if it exists, check if the namespace is the same.
+    ///
+    /// # Arguments:
+    ///
+    /// * `workload_id`: The workload id to get
+    /// * `namespace`: The namespace of the workload
+    ///
+    /// # Returns:
+    ///
+    /// A Result<String, WorkloadError>
 
-#[patch("/{workload_id}")]
-pub async fn patch_workload(
-    workload_id: web::Path<String>,
-    body: web::Json<WorkloadInfo>,
-) -> impl Responder {
-    let workload_info = body.into_inner();
-    service::update_workload(&workload_id, &workload_info);
-    HttpResponse::Ok().body(format!("patch_workload {}", workload_id))
-}
+    pub async fn workload(
+        params: web::Path<(String, String)>,
+        data: web::Data<ActixAppState>,
+    ) -> impl Responder {
+        let (namespace, workload_id) = params.into_inner();
+
+        let mut workload_service = WorkloadService::new_or_http_error(&data.etcd_address)
+            .await
+            .unwrap();
+
+        match workload_service
+            .get_workload(&workload_id, &namespace)
+            .await
+        {
+            Ok(workload) => workload.to_http(),
+            Err(e) => e.to_http(),
+        }
```

Comment:
This is `map_or_else(e e.to_http(), w w.to_http())?`

Reply:
this looks better however HttpResponse from actix does not implement the Try feature to use `?`

Reply:
You actually don't need the `?`. The `map_or_else`call should return `HttpResponse`, which implements `Responder`afaiu.

Reply:
yes my bad